lisa/runner.py

```diff
@@ -282,52 +291,41 @@ def _start_loop(self) -> None:
+                        if runner.is_done:
+                            # remove fully completed runner.
+                            runner.close()
+                            remaining_runners.remove(runner)
                             self._log.debug(
-                                f"fetched task from {runner.id}: '{task_message}'"
+                                f"runner '{runner.id}' is done, "
+                                f"remaining runners {[x.id for x in remaining_runners]}"
                             )
-                            task_manager.submit_task(task)
+
+                    if task_manager.has_idle_worker():
+                        if has_more_runner:
+                            # Add new runner if idle workers are available.
+                            try:
+                                remaining_runners.append(next(runner_iterator))
+                            except StopIteration:
+                                has_more_runner = False
                         else:
-                            # current runner may not be done, but it doesn't
-                            # have task temporarily. The root runner can start
-                            # tasks from next runner.
-                            break
-                        if not task_manager.has_idle_worker():
-                            has_idle_worker = False
-                            break
-                    if runner.is_done:
-                        # remove fully completed runner.
-                        runner.close()
-                        remaining_runners.remove(runner)
-                        self._log.debug(
-                            f"runner '{runner.id}' is done, "
-                            f"remaining runners {[x.id for x in remaining_runners]}"
-                        )
-                        break
-                    if not has_idle_worker:
-                        # uses the result from previous runner, because the
-                        # worker status may be changed after first runner
-                        # finished. Evne in this case, it should try result from
-                        # previous runner firstly in next run.
-                        break
+                            # Reduce CPU utilization from infinite loop when idle
+                            # workers are present but no task to run.
+                            time.sleep(1)
 
-                while (
-                    len(remaining_runners) < self._max_concurrency and has_more_runner
-                ):
-                    # Fetch runners, if runner count is smaller than concurrency
-                    # count. It makes sure all concurrency can run.
-                    try:
-                        remaining_runners.append(next(runner_iterator))
-                    except StopIteration:
-                        has_more_runner = False
+                    if not remaining_runners and not has_more_runner:
```

Comment:
This check can be merged with the top level while by adding `or has_more_runner`